File: backend/internal/models/sale.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"
)

// SaleStatus represents the status of a sale.
type SaleStatus string

const (
	SaleStatusDraft     SaleStatus = "draft"
	SaleStatusCompleted SaleStatus = "completed"
	SaleStatusVoided    SaleStatus = "voided"
)

// PaymentStatus represents the payment status of a sale.
type PaymentStatus string

const (
	PaymentStatusPending  PaymentStatus = "pending"
	PaymentStatusPartial  PaymentStatus = "partial"
	PaymentStatusPaid     PaymentStatus = "paid"
	PaymentStatusRefunded PaymentStatus = "refunded"
	PaymentStatusVoided   PaymentStatus = "voided"
)

// Sale represents a sales transaction.
type Sale struct {
	ID        uuid.UUID `json:"id"`
	InvoiceNo string    `json:"invoice_no"`

	Subtotal       decimal.Decimal `json:"subtotal"`
	TaxAmount      decimal.Decimal `json:"tax_amount"`
	DiscountAmount decimal.Decimal `json:"discount_amount"`
	TotalAmount    decimal.Decimal `json:"total_amount"`
	ItemCount      int             `json:"item_count"`

	PaymentStatus PaymentStatus   `json:"payment_status"`
	AmountPaid    decimal.Decimal `json:"amount_paid"`
	ChangeAmount  decimal.Decimal `json:"change_amount"`

	DiscountType   *string          `json:"discount_type,omitempty"`
	DiscountValue  *decimal.Decimal `json:"discount_value,omitempty"`
	DiscountReason *string          `json:"discount_reason,omitempty"`

	EmployeeID uuid.UUID  `json:"employee_id"`
	ShiftID    *uuid.UUID `json:"shift_id,omitempty"`

	Status     SaleStatus `json:"status"`
	VoidedAt   *time.Time `json:"voided_at,omitempty"`
	VoidedBy   *uuid.UUID `json:"voided_by,omitempty"`
	VoidReason *string    `json:"void_reason,omitempty"`

	Notes *string `json:"notes,omitempty"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`

	Items        []SaleItem `json:"items,omitempty"`
	Payments     []Payment  `json:"payments,omitempty"`
	EmployeeName *string    `json:"employee_name,omitempty"`
}

// SaleItem represents a line item in a sale.
type SaleItem struct {
	ID     uuid.UUID `json:"id"`
	SaleID uuid.UUID `json:"sale_id"`

	ProductID      uuid.UUID `json:"product_id"`
	ProductName    string    `json:"product_name"`
	ProductSKU     *string   `json:"product_sku,omitempty"`
	ProductBarcode *string   `json:"product_barcode,omitempty"`

	Quantity  decimal.Decimal `json:"quantity"`
	UnitPrice decimal.Decimal `json:"unit_price"`
	CostPrice decimal.Decimal `json:"cost_price"`

	DiscountType   *string         `json:"discount_type,omitempty"`
	DiscountValue  decimal.Decimal `json:"discount_value"`
	DiscountAmount decimal.Decimal `json:"discount_amount"`

	TaxRate   decimal.Decimal `json:"tax_rate"`
	TaxAmount decimal.Decimal `json:"tax_amount"`
	Subtotal  decimal.Decimal `json:"subtotal"`
	Total     decimal.Decimal `json:"total"`

	CreatedAt time.Time `json:"created_at"`
}

File: frontend/src/types/sale.ts
import type { Product } from "./catalog";

export type SaleStatus = "completed" | "voided" | "pending";
export type PaymentMethod = "cash" | "card" | "qris" | "transfer";

export interface SaleItem {
  id: string;
  product_id: string;
  product_name: string;
  product_sku?: string;
  product_barcode?: string;
  quantity: string;
  unit_price: string;
  discount_amount: string;
  tax_rate: string;
  tax_amount: string;
  subtotal: string;
  total: string;
}

export interface Sale {
  id: string;
  invoice_no: string;
  employee_id: string;
  employee_name?: string;
  shift_id?: string;
  status: SaleStatus;
  subtotal: string;
  tax_amount: string;
  discount_amount: string;
  total_amount: string;
  amount_paid: string;
  change_amount: string;
  item_count: number;
  payment_status: string;
  notes?: string;
  void_reason?: string;
  created_at: string;
  updated_at: string;
  items?: SaleItem[];
  payments?: SalePayment[];
}

export interface SalePayment {
  id?: string;
  payment_method: PaymentMethod;
  amount: string;
  amount_tendered?: string;
  change_given?: string;
  reference_no?: string;
  notes?: string;
}

export interface CreateSaleItem {
  product_id: string;
  quantity: string;
  unit_price: string;
  discount_value?: string;
  discount_amount?: string;
}

export interface CreatePayment {
  payment_method: PaymentMethod;
  amount: string;
  amount_tendered?: string;
  change_given?: string;
  reference_no?: string;
  notes?: string;
}

export interface CreateSaleRequest {
  items: CreateSaleItem[];
  payments: CreatePayment[];
  discount_type?: string;
  discount_value?: string;
  discount_reason?: string;
  notes?: string;
}

export interface CartItem {
  product: Product;
  quantity: number;
}

export interface Cart {
  items: CartItem[];
  subtotal: number;
  discount: number;
  tax: number;
  total: number;
}
